site/src/contexts/ProxyContext.tsx

```diff
@@ -0,0 +1,211 @@
+import { useQuery } from "@tanstack/react-query"
+import { getApplicationsHost, getWorkspaceProxies } from "api/api"
+import { Region } from "api/typesGenerated"
+import { useDashboard } from "components/Dashboard/DashboardProvider"
+import {
+  createContext,
+  FC,
+  PropsWithChildren,
+  useContext,
+  useState,
+} from "react"
+
+interface ProxyContextValue {
+  proxy: PreferredProxy
+  isLoading: boolean
+  error?: Error | unknown
+  setProxy: (regions: Region[], selectedRegion: Region | undefined) => void
+}
+
+interface PreferredProxy {
+  // SelectedRegion is the region the user has selected.
+  // Do not use the fields 'path_app_url' or 'wildcard_hostname' from this
+  // object. Use the preferred fields.
+  selectedRegion: Region | undefined
+  // PreferredPathAppURL is the URL of the proxy or it is the empty string
+  // to indicate using relative paths. To add a path to this:
+  //  PreferredPathAppURL + "/path/to/app"
+  preferredPathAppURL: string
+  // PreferredWildcardHostname is a hostname that includes a wildcard.
+  preferredWildcardHostname: string
+}
+
+export const ProxyContext = createContext<ProxyContextValue | undefined>(
+  undefined,
+)
+
+/**
+ * ProxyProvider interacts with local storage to indicate the preferred workspace proxy.
+ */
+export const ProxyProvider: FC<PropsWithChildren> = ({ children }) => {
+  // Try to load the preferred proxy from local storage.
+  let savedProxy = loadPreferredProxy()
+  if (!savedProxy) {
+    savedProxy = getURLs([])
+  }
+
+  // The initial state is no regions and no selected region.
+  const [proxy, setProxy] = useState<PreferredProxy>(savedProxy)
+  const setAndSaveProxy = (
+    regions: Region[],
+    selectedRegion: Region | undefined,
+  ) => {
+    const preferred = getURLs(regions, selectedRegion)
+    // Save to local storage to persist the user's preference across reloads
+    // and other tabs.
+    savePreferredProxy(preferred)
+    // Set the state for the current context.
+    setProxy(preferred)
+  }
+
+  const queryKey = ["get-regions"]
+  const { error: regionsError, isLoading: regionsLoading } = useQuery({
+    queryKey,
+    queryFn: getWorkspaceProxies,
```

Comment:
Because this Provider wraps the whole app, we are making this call everywhere, including when we don't have this feature enabled which results in an automatic 500 console error:
![Screenshot 2023-04-27 at 7 47 06 PM](https://user-images.githubusercontent.com/19142439/235013656-057b2583-c789-4b32-848f-6e563c05b429.png)
We should figure out another strategy here (maybe returning early after an entitlement check) so that we don't have persistent console errors.

Reply:
Did you pull the latest?

I made a change so the `/regions` endpoint is **always available** to fix this. Even if the feature is not enabled.